determined-ai/determined
# feat: refactor compareTrials to support time-series [WEB-999]

## Description

feat: refactor compareTrials to support time-series [WEB-999]

Implements the `time-series` API via refactoring the `compareTrials` API. Discussions with the team led to the decisions that the proposed `time-series` API and current `compareTrials` API contained enough overlap that refactoring the `compareTrials` API to support `time-series` was more favorable. 

Two initiatives from the `unified metrics` project have also been implemented in this PR. 

1. We want to stop downsampling using `lttb` in favor of random downsampling, therefore `lttb` has now been removed.
2. Previously we returned metrics in the following way: 

```
{
    epochs: [<metric_value>, <epoch_value>],
    time: [<metric_value>, <time_value>],
    batches: [<metric_value>, <batches_value>],
}
```
where each array in the previous example was downsampled independently. This has been refactored now that all points are downsampled together.

```
{
 "value": <metric_value>
 "epochs": <epoch_value>,
 "time": <time_value>,
  "batches:" <batches_value>
}
```

API usage expects that callers will define a `name` for the metric or column to query by, as well as one singular filter range. For example, the API can be called with `batches` and a specified `integerRange` to receive a metric between a range of batches, or `time` and a `timeRange` to receive metrics between two times. 

 




## Test Plan

Use the swagger page to ensure the `time-series` API works correctly

1.  Send a request with no parameters specified and ensure you receive an empty list such as `trials: []`
2. Specify at least one valid `metricId` such as "validation.validation_error" and `trialId` and ensure that you receive at most 200 data points back. The datapoints in the `metrics` data should all contain "batches", "value", and "time".
3. Specify both a valid "training" and "validation" `metricId` such as "validation.validation_error" and  "training.train_error" and `trialId` and ensure that you receive at most 200 data points back. The datapoints in the different types of `metrics` data should all contain "batches", "value", and "time".
4. Specify a `maxDatapoints` value and ensure at most  that number of data points are returned per metric.
5. Specify `batches` as the  `timeSeriesFilter.name` and ensure the results in (2) and (3) are the same. This confirms that batches is the default sort and order column.
6. Only specify an timeSeriesFilter.IntegerRange.lt, ensure that the largest value of `batches` is less than that number. 
7. Only specify an timeSeriesFilter.IntegerRange.lte, ensure that the largest value of `batches` is less than or equal to that number. 
8. Only specify an timeSeriesFilter.IntegerRange.gt, ensure that the smallest value of `batches` is greater than that number. 
9. Only specify an  timeSeriesFilter.IntegerRange.gte, ensure that the smallest value of `batches` is greater than or equal to that number. 
10. Specify an IntegerRange.gt value, then specify a IntegerRange.lt that is lower, ensure you receive an error message.
11. Specify an IntegerRange.gt value, then specify a IntegerRange.lte that is lower, ensure you receive an error message.
12. Repeat (9) and (10) instead using `IntegerRange.gte`
13.  Repeat (9-11) but specify both `IntegerRange.lte` and `IntegerRange.lt` with one being lower than the greater than value. Ensure that you receive an error message.
14. Specify a rangetype of `rangeType Date`. Only specify an timeRange.lt, ensure that the largest value of `time` is less than that number. 
15. Repeat (10-14) with timeSeriesFilter.doubleRange instead
16. Only specify a timeRange.lte, ensure that the largest value of `time` is less than or equal to that number. 
17. Only specify a timeRange.gt, ensure that the smallest value of `time` is greater than that number. 
18. Only specify a timeRange.gte, ensure that the smallest value of `time` is greater than or equal to that number. 
19. Specify a timeRange.gt value, then specify a timeRange.lt that is lower, ensure you receive an error message.
20. Specify a timeRange.gt value, then specify a timeRange.lte that is lower, ensure you receive an error message.
21. Repeat (9) and (10) instead using `timeRange.gte`
22.  Repeat (9-11) but specify both `timeRange.lte` and `timeRange.lt` with one being lower than the greater than value. Ensure that you receive an error message.
23. Send a request with both `metricNames` and `metricIds` specified. Ensure that you receive an error.
24. Send a request with at least one value for at least two different ranges, ensure that you receive an error.
25. On the experiment overview pages, both with and without `f_chart=on` ensure that the graphs and datapoints look nearly identical to how they were shown previously. Since downsampling logic has changed there might be slight variations in how charts look. 





## Commentary (optional)

The vast majority of the refactoring has been implemented as `additive` changes in order to preserve current `compareTrials` functionality. However, if we feel it is better to additionally change some of the current `compareTrials` logic that can be done. 






## Checklist

~- [ ] User-facing API changes need the "User-facing API Change" label.~
~- [ ] Release notes should be added as a separate file under `docs/release-notes/`.~
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
~- [ ] Licenses should be included for new code which was copied and/or modified from any external code.~

## Ticket


WEB-999




[WEB-999]: https://hpe-aiatscale.atlassian.net/browse/WEB-999?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
